File: js/news.js
/**
 * Tech & World News Aggregator
 */
class NewsManager {
  constructor() {
    this.articles = [];
    this.currentCat = 'tech';
  }

  init() {
    this.bindEvents();
    this.fetchNews('tech');
  }

  bindEvents() {
    const pills = document.querySelectorAll('.news-categories .cat-pill');
    pills.forEach(pill => {
      pill.addEventListener('click', () => {
        pills.forEach(p => p.classList.remove('active'));
        pill.classList.add('active');
        this.fetchNews(pill.dataset.cat);
      });
    });

    const searchInput = document.getElementById('newsFilterInput');
    if (searchInput) {
      searchInput.addEventListener('input', (e) => {
        this.renderArticles(e.target.value.trim().toLowerCase());
      });
    }
  }

  async fetchNews(category) {
    this.currentCat = category;
    const listEl = document.getElementById('newsList');
    if (listEl) listEl.innerHTML = '<div class="news-item-skeleton">Fetching latest headlines...</div>';

    try {
      // 1. Try local server daemon proxy endpoint
      let res = await fetch('/api/hn').catch(() => null);
      if (res && res.ok) {
        this.articles = await res.json();
      } else {
        // 2. Direct Hacker News API fallback
        const hnRes = await fetch('https://hacker-news.firebaseio.com/v0/topstories.json');
        const ids = (await hnRes.json()).slice(0, 8);
        this.articles = await Promise.all(
          ids.map(async id => {
            const itemRes = await fetch(`https://hacker-news.firebaseio.com/v0/item/${id}.json`);
            const item = await itemRes.json();
            return {
              title: item.title,
              url: item.url || `https://news.ycombinator.com/item?id=${id}`,
              by: item.by || 'hn',
              score: item.score || 0
            };
          })
        );
      }
    } catch (e) {
      this.articles = [
        { title: "Apple Unveils macOS Sequoia with Seamless iPhone Mirroring", url: "https://apple.com", by: "tech", score: 420 },
        { title: "Node.js Releases High-Performance Native SQLite & Web API Support", url: "https://nodejs.org", by: "dev", score: 310 },
        { title: "Python 3.13 Introduces Experimental Free-Threaded GIL Removal", url: "https://python.org", by: "dev", score: 280 },
        { title: "Vite 6.0 Alpha Benchmark: 10x Faster Cold Start Initialization", url: "https://vitejs.dev", by: "tech", score: 195 }
      ];
    }

    this.renderArticles();
  }

  renderArticles(query = '') {
    const listEl = document.getElementById('newsList');
    if (!listEl) return;

    let filtered = this.articles;
    if (query) {
      filtered = this.articles.filter(a => a.title.toLowerCase().includes(query));
    }

    if (filtered.length === 0) {
      listEl.innerHTML = '<div class="news-item-skeleton">No matching news items found.</div>';
      return;
    }

    listEl.innerHTML = '';
    filtered.forEach(art => {
      const item = document.createElement('div');
      item.className = 'news-item';
      item.innerHTML = `
        <a href="${art.url}" target="_blank" rel="noopener noreferrer">⚡ ${art.title}</a>
        <span style="color:var(--text-muted);font-size:10px;white-space:nowrap;">▲ ${art.score || 0}</span>
      `;
      listEl.appendChild(item);
    });
  }
}

window.newsManager = new NewsManager();

File: avocado/news.py
"""
Terminal News Reader (Hacker News & Tech Headlines)
"""
import urllib.request
import json

def get_top_news(category="tech", count=5):
    try:
        url = "https://hacker-news.firebaseio.com/v0/topstories.json"
        req = urllib.request.urlopen(url, timeout=3)
        ids = json.loads(req.read().decode('utf-8'))[:count]
        
        items = []
        for i_id in ids:
            try:
                item_url = f"https://hacker-news.firebaseio.com/v0/item/{i_id}.json"
                i_req = urllib.request.urlopen(item_url, timeout=2)
                item_data = json.loads(i_req.read().decode('utf-8'))
                if item_data and "title" in item_data:
                    items.append({
                        "title": item_data.get("title"),
                        "url": item_data.get("url", f"https://news.ycombinator.com/item?id={i_id}"),
                        "score": item_data.get("score", 0)
                    })
            except Exception:
                continue
        return items
    except Exception:
        return [
            {"title": "Apple Unveils macOS Sequoia with Seamless iPhone Mirroring", "url": "https://apple.com", "score": 420},
            {"title": "Node.js Releases High-Performance Native SQLite Support", "url": "https://nodejs.org", "score": 310},
            {"title": "Python 3.13 Introduces Experimental Free-Threaded GIL", "url": "https://python.org", "score": 280},
            {"title": "Vite 6.0 Benchmark: 10x Faster Cold Start Initialization", "url": "https://vitejs.dev", "score": 195}
        ]
